cosmos/operators/virtualenv.py

```diff
@@ -102,11 +117,110 @@ def run_subprocess(self, command: list[str], env: dict[str, str], cwd: str) -> F
         )
         return subprocess_result
 
+    def clean_dir_if_temporary(self) -> None:
+        """
+        Delete the virtualenv directory if it is temporary.
+        """
+        if self.is_virtualenv_dir_temporary and self.virtualenv_dir and self.virtualenv_dir.exists():
+            self.log.info(f"Deleting the Python virtualenv {self.virtualenv_dir}")
+            shutil.rmtree(str(self.virtualenv_dir), ignore_errors=True)
+
     def execute(self, context: Context) -> None:
-        output = super().execute(context)
-        if self._venv_tmp_dir:
-            self._venv_tmp_dir.cleanup()
-        logger.info(output)
+        try:
+            super().execute(context)
+        finally:
+            self.clean_dir_if_temporary()
+
+    def on_kill(self) -> None:
+        self.clean_dir_if_temporary()
+
+    def prepare_virtualenv(self) -> str:
+        self.log.info(f"Creating or updating the virtualenv at `{self.virtualenv_dir}")
+        py_bin = prepare_virtualenv(
+            venv_directory=str(self.virtualenv_dir),
+            python_bin=PY_INTERPRETER,
+            system_site_packages=self.py_system_site_packages,
+            requirements=self.py_requirements,
+            pip_install_options=self.pip_install_options,
+        )
+        return py_bin
+
+    def _get_or_create_venv_py_interpreter(self) -> str:
+        """Helper method that parses virtual env configuration
+        and returns a DBT binary within the resulting virtualenv"""
+
+        # No virtualenv_dir set, so create a temporary virtualenv
+        if self.virtualenv_dir is None or self.is_virtualenv_dir_temporary:
+            self.log.info("Creating temporary virtualenv")
+            with TemporaryDirectory(prefix="cosmos-venv") as tempdir:
+                self.virtualenv_dir = Path(tempdir)
+                py_bin = self.prepare_virtualenv()
+            return py_bin
+
+        # Use a reusable virtualenv
+        self.log.info(f"Checking if the virtualenv lock {str(self._lock_file)} exists")
+        while not self._is_lock_available() and self.max_retries_lock:
+            logger.info("Waiting for virtualenv lock to be released")
+            time.sleep(1)
+            self.max_retries_lock -= 1
+
```

Comment:
just curious about how often we need to wait for the lock to be released. 
Do we have any data on how performance is affected if 100 parallel tasks try to use this virtual environment?

Reply:
We don't have this data right now, and I agree it would be great if we could get this. @pankajastro , could you help test this, please?

Reply:
We spoke to @pankajastro and we decided to release this feature to collect feedback from the community, and get support from others on testing this